src/clib/pio_spmd.c

```diff
@@ -305,108 +251,97 @@ int pio_swapm(void *sndbuf, int *sndlths, int *sdispls, MPI_Datatype *stypes,
 #ifdef DEBUG
         int totalrecv = 0;
         int totalsend = 0;
-        for (int i = 0; i < nprocs; i++)
-	{
-            totalsend += sndlths[i];
-            totalrecv += rcvlths[i];
+        for (int i = 0; i < ntasks; i++)
+        {
+            totalsend += sendcounts[i];
+            totalrecv += recvcounts[i];
         }
-        printf("%s %d totalsend %d totalrecv %d \n",__FILE__,__LINE__,totalsend,totalrecv);
+        printf("%s %d totalsend %d totalrecv %d \n", __FILE__, __LINE__, totalsend,
+               totalrecv);
 #endif
 #ifdef OPEN_MPI
-        for (int i = 0; i < nprocs; i++)
-	{
-            if (stypes[i] == MPI_DATATYPE_NULL)
-                stypes[i] = MPI_CHAR;
-            if (rtypes[i] == MPI_DATATYPE_NULL)
-                rtypes[i] = MPI_CHAR;
+        /* For some reason OPEN_MPI has problems with MPI_CHAR ??? */
+        for (int i = 0; i < ntasks; i++)
+        {
+            if (sendtypes[i] == MPI_DATATYPE_NULL)
+                sendtypes[i] = MPI_CHAR;
+            if (recvtypes[i] == MPI_DATATYPE_NULL)
+                recvtypes[i] = MPI_CHAR;
         }
 #endif
-        CheckMPIReturn(MPI_Alltoallw(sndbuf, sndlths, sdispls, stypes, rcvbuf,
-				     rcvlths, rdispls, rtypes, comm),__FILE__,__LINE__);
+
+        /* Call the MPI alltoall without flow control. */
+        CheckMPIReturn(MPI_Alltoallw(sendbuf, sendcounts, sdispls, sendtypes, recvbuf,
+                                     recvcounts, rdispls, recvtypes, comm), __FILE__, __LINE__);
 
 #ifdef OPEN_MPI
-        for (int i = 0; i < nprocs; i++)
-	{
-            if (stypes[i] == MPI_CHAR)
-                stypes[i] = MPI_DATATYPE_NULL;
-            if (rtypes[i] == MPI_CHAR)
-                rtypes[i] = MPI_DATATYPE_NULL;
+        /* For some reason OPEN_MPI has problems with MPI_CHAR ??? */
+        for (int i = 0; i < ntasks; i++)
+        {
+            if (sendtypes[i] == MPI_CHAR)
+                sendtypes[i] = MPI_DATATYPE_NULL;
+            if (recvtypes[i] == MPI_CHAR)
+                recvtypes[i] = MPI_DATATYPE_NULL;
         }
 #endif
         return PIO_NOERR;
     }
 
-    int tag;
-    int offset_t;
-    int ierr;
-    MPI_Status status;
-    int steps;
-    int istep;
-    int rstep;
-    int p;
-    int maxreq;
-    int maxreqh;
-    int hs;
-    int cnt;
-    void *ptr;
-    MPI_Request rcvids[nprocs];
+    /* ??? */
+    offset_t = ntasks;
 
-    offset_t = nprocs;
-    // send to self
-    if (sndlths[mytask] > 0)
+    /* Send to self. */
+    if (sendcounts[my_rank] > 0)
     {
         void *sptr, *rptr;
         int extent, lb;
-        tag = mytask + offset_t;
-        sptr = (void *)((char *) sndbuf + sdispls[mytask]);
-        rptr = (void *)((char *) rcvbuf + rdispls[mytask]);
+        tag = my_rank + offset_t;
+        sptr = (char *)sendbuf + sdispls[my_rank];
```

Comment:
The char was already there (and is needed to make the pointer arithmetic work). The cast of the result to void * is not necessary. It will happen anyway when sptr (which is a void *) gets the char * which is the result of the expression. It's part of the definition of void - we don't need to do it in our code.